### PR TITLE
review miller rabin

### PR DESCRIPTION
**Pull request by thor314**

### CODE REVIEW

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 1]
```diff
@@ -1,4 +1,5 @@
 use std::vec;
+// tk: cargo clippy --fix  would solve a lot of linting errors, clippy is your friend
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:2]: I'm actually using clippy already (it's run for each commit with GH action):
https://github.com/ArnaudBrousseau/crypto-study-group/blob/f7bcff60f5294241c4975c2cf5b2c569fdee6abd/.github/workflows/ci.yml#L27

I thought CI would be failing but apparently not. I get _some_ of the fixes you pointed out but not all of them when I run `cargo clippy --fix` on my machine. Are you using nightly or some sort of special config on your side?

Anyway, fixed what I found locally: https://github.com/ArnaudBrousseau/crypto-study-group/commit/e265038f268fe86461e640e68830fa06dcd05ef7

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 2]
```diff
@@ -9,7 +10,7 @@ pub enum MillerRabinError {
     #[error("This function only supports up to 1,024 bytes of input. Got {0}.")]
     TooManyBytes(usize),
     #[error("The Miller-Rabin test only allows testing numbers > 3. Found n <= 3")]
-    LowInteger(),
+    LowInteger(), // tk: don't need parens here
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:13]: Thanks for the tip! https://github.com/ArnaudBrousseau/crypto-study-group/commit/ca4c2550a1e6746f1a9eda45673d3479553cc396

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 3]
```diff
@@ -18,6 +19,8 @@ pub enum MillerRabinError {
     MulModOperandTooHigh(),
     #[error("Exponentiation modulo p requires operand to be < p")]
     PowMulOperandTooHigh(),
+    #[error("misc")]
+    Misc,
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:23]: Mind expanding on this? Is that a convention?

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 4]
```diff
@@ -26,12 +29,26 @@ pub enum MillerRabinError {
 /// Under the hood, we use crypto-bigint's U8192.
 /// Hence this function errors if `bytes` has more than 1,024 u8 elements.
 pub fn miller_rabin(bytes: Vec<u8>) -> Result<bool, MillerRabinError> {
-    const REQUIRED_BYTE_LENGTH_FOR_U8192: usize = 1024;
+    // tk: polite to take bytes by reference, don't take ownership:
+    // pub fn miller_rabin(bytes: &[u8]) -> Result<bool, MillerRabinError> {
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:33]: Ah. Good to know! Switched this up: https://github.com/ArnaudBrousseau/crypto-study-group/commit/8ef1942157a71252b09d8aaa0a7a04ebe41de1ef

It had the nice side-effect of making tests less obnoxious to write (no more `to_vec`)

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 5]
```diff
@@ -26,12 +29,26 @@ pub enum MillerRabinError {
...
+    // Also, it would be nice to return the witness if one is found, rather than just true, eg:
+    // pub fn miller_rabin(bytes: &[u8]) -> Result<Option<U8192>, MillerRabinError> {
+    const REQUIRED_BYTE_LENGTH_FOR_U8192: usize = 1024; // tk: by convention, consts live at top-level, if want lazy execution, use a lazy once cell static
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:36]: Good call 👍 

https://github.com/ArnaudBrousseau/crypto-study-group/commit/d6adc671a48ac09fa92b44d91d70b70d2e8c22fa

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 6]
```diff
@@ -26,12 +29,26 @@ pub enum MillerRabinError {
 /// Under the hood, we use crypto-bigint's U8192.
 /// Hence this function errors if `bytes` has more than 1,024 u8 elements.
 pub fn miller_rabin(bytes: Vec<u8>) -> Result<bool, MillerRabinError> {
-    const REQUIRED_BYTE_LENGTH_FOR_U8192: usize = 1024;
+    // tk: polite to take bytes by reference, don't take ownership:
+    // pub fn miller_rabin(bytes: &[u8]) -> Result<bool, MillerRabinError> {
+    // Also, it would be nice to return the witness if one is found, rather than just true, eg:
+    // pub fn miller_rabin(bytes: &[u8]) -> Result<Option<U8192>, MillerRabinError> {
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:35]: I don't really want to leak internal types (U8192) so I'm thinking I can return bytes? Or a hex string?

I'm also wondering if the interface would be clearer with 3 values:
* `bool` to indicate whether the test is successful
* `Option<&str>` (witness)
We'd have a return type of `Result<(bool, Option<&str>), MillerRabinError>`. Quite a mouthful.

Maybe there's an opportunity for a tiny struct actually? Something like `Result<MillerRabinResult, MillerRabinError>` (and `MillerRabinResult` contains a success bool + a option for a witness). That might be easier for callers to reason about what they're getting back.

Curious to hear your thoughts on one vs the other.

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 7]
```diff
@@ -26,12 +29,26 @@ pub enum MillerRabinError {
...
+    const REQUIRED_BYTE_LENGTH_FOR_U8192: usize = 1024; // tk: by convention, consts live at top-level, if want lazy execution, use a lazy once cell static
 
     if bytes.len() > REQUIRED_BYTE_LENGTH_FOR_U8192 {
+        // tk: option to use ? operator:
+        // Err(MillerRabinError::TooManyBytes(bytes.len()))?;
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:40]: 👍 

https://github.com/ArnaudBrousseau/crypto-study-group/commit/78965440031bd708424e71a78674092fc1adc256

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 8]
```diff
@@ -26,12 +29,26 @@ pub enum MillerRabinError {
...
         return Err(MillerRabinError::TooManyBytes(bytes.len()));
     }
 
+    // tk: option to not allocate vector to the heap here, array on the stack more efficient
+    // let arr: [u8; REQUIRED_BYTE_LENGTH_FOR_U8192] = {
+    //     let mut arr = [0u8; REQUIRED_BYTE_LENGTH_FOR_U8192];
+    //     arr[0..bytes.len()].copy_from_slice(&bytes);
+    //     arr
+    // };
+    // let n = U8192::from_be_slice(&arr);
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:50]: Nice tip 🙌 

https://github.com/ArnaudBrousseau/crypto-study-group/commit/8f1bc061a56b316146378fd453aa8d6da092d89f

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 9]
```diff
@@ -87,6 +111,47 @@ pub fn miller_rabin(bytes: Vec<u8>) -> Result<bool, MillerRabinError> {
         }
     }
     Ok(true)
+
+    // // tk: imperative vs functional style depends, but when possible, functional is more commonly used, often more
+    // // readable, and easier to debug. Your algorithm doesn't actually use your iterator value, so is wasted CPU.
+    // // Admittedly, satisfying the type system with functional algorithms can be a bit tricky, can take longer to write.
+    // //
+    // // tk: functional equiv.
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:119]: Thanks for taking the time to write this up! I'll digest this at my own pace; lots of good rusty things in there!

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 10]
```diff
@@ -52,18 +69,25 @@ pub fn miller_rabin(bytes: Vec<u8>) -> Result<bool, MillerRabinError> {
...
     }
 
+    // tk: production code should comment why this is always safe to unwrap, or else return an error
     // Compute (s, t) such that s is odd and s.2^t = n-1
     let n_minus_one = n.checked_sub(&U8192::ONE).unwrap();
     let mut s = n_minus_one;
     let mut t = 0;
+    // tk: don't allocate two repeatedly
+    // let two = &U8192::from(2u8);
+    // tk: unhandled edge case, where n = 2.
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:79]: I'm checking for n <= 3 above:
https://github.com/ArnaudBrousseau/crypto-study-group/blob/8f1bc061a56b316146378fd453aa8d6da092d89f/2023-01-13/miller_rabin/src/lib.rs#L42-L44

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 11]
```diff
@@ -52,18 +69,25 @@ pub fn miller_rabin(bytes: Vec<u8>) -> Result<bool, MillerRabinError> {
         return Err(MillerRabinError::EvenInteger());
     }
 
+    // tk: production code should comment why this is always safe to unwrap, or else return an error
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:72]: Makes sense! https://github.com/ArnaudBrousseau/crypto-study-group/commit/c739f3f327cc1f685e228579c6df06a52788e459

File: 2023-01-13/miller_rabin/src/lib.rs [hunk 12]
```diff
@@ -52,18 +69,25 @@ pub fn miller_rabin(bytes: Vec<u8>) -> Result<bool, MillerRabinError> {
...
         s = s.checked_div(&U8192::from(2u8)).unwrap();
+        // s = s.checked_div(&two).unwrap();
         t += 1;
     }
 
     let mut k = 0;
+    // tk: unnecessary copy
     let limit = n_minus_one;
     let bit_size = bit_len(&n);
+    // tk: 128 -> option to use a const for self-documenting var name
```
Comment (ArnaudBrousseau) [inline 2023-01-13/miller_rabin/src/lib.rs:90]: Nice catch! https://github.com/ArnaudBrousseau/crypto-study-group/commit/1902f44775b23ab5b44f1ca6d81c7957a6b82771

**Review by ArnaudBrousseau — COMMENTED** [review]
Thank you so much for the thorough review @thor314 🙇 

I've linked the individual commits (where fixes live), but you can see the aggregate here as well: https://github.com/ArnaudBrousseau/crypto-study-group/compare/f7bcff60..8ef1942

File: 2023-01-13/miller_rabin/src/lib.rs (hunk 1 above)
Comment (thor314) [reply 2023-01-13/miller_rabin/src/lib.rs:2]: huh, you might be running another rust toolchain? I'm on 1.68 nightly, you may be on stable

Comment (thor314) [reply 2023-01-13/miller_rabin/src/lib.rs:2]: the only changes made, besides my comments, were inserted by `cargo clippy --fix --dirty-something`

File: 2023-01-13/miller_rabin/src/lib.rs (hunk 3 above)
Comment (thor314) [reply 2023-01-13/miller_rabin/src/lib.rs:23]: i was lazy and needed a return type

**Comment by Liamdruk** [comment]
Hello, if you were into any investment or crypto scam. I will recommend and advice to reach out to The Recovery Agency, with no upfront fees, I recently got my money back.
Company Verified Telegram;https://t.me/TheRecovery_Agency

**Pull request closed without merging**